### DISCUSSION TITLE
Custom Playbook

### DISCUSSION THREAD

**Discussion by Mav1814** [Security-Onion-Solutions/securityonion#11499] [2.4]
### Version

Other (please provide detail below)

### Installation Method

Security Onion ISO image

### Description

configuration

### Installation Type

Distributed

### Location

on-prem with Internet access

### Hardware Specs

Exceeds minimum requirements

### CPU

4

### RAM

32

### Storage for /

104G

### Storage for /nsm

206G

### Network Traffic Collection

tap

### Network Traffic Speeds

1Gbps to 10Gbps

### Status

Yes, all services on all nodes are running OK

### Salt Status

Yes, there are salt failures (please provide detail below)

### Logs

No, there are no additional clues

### Detail

Hello again guys!

I've been trying to create a playbook to detect brute force attempts on logins. 
I don't get any errors when converting sigma to ElastAlert Config, it shows the conversion. 
After creating the playbook and making it active, what are the steps to take? How do I know it's working or if it's failing?
I've already made a few failed attempts on one of my endpoints and no alerts have appeared.

thank you in advance
cheers

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to Security Onion 2.4.

**Comment by Taishinto** [comment]
I like this idea, and may be interested as well.  
Would you mind sharing the code - I'd first look at what/how the alert is triggered and what the threshold is for that trigger.

**Reply by Mav1814** [reply]
Hello!

I've been trying to develop, I don't have a working sigma code yet. I've been trying to create it using kibana queries and adapt it for the playbook.

My big question is, how do I know that my created playbook will be active and working (after the active option)?

cheers

**Reply by Mav1814** [reply]
hi! 
anyone? 
 How can i test my playbook is working? i've created 1 playbook, i tested the output of the query on kibana and it worked. But when i try to trigger, i dont receive any alert. :/

**Reply by keis3cker** [reply]
you should be able  to see under /opt/so/rules/elastalert/playbook if your playook has been imported into elastalert.

if not, you can execute so-playbook-sync, this should import your active playbooks into elastalert. maybe there's a syntax error in your rule and can't be imported to elastalert. The output of so-playbook-sync is helpful in that case

**Reply by Mav1814** [reply]
Hello, 

In fact, the playbook I created is imported. The only doubt is that it doesn't have a "play_id".  It shows as ".yaml" . 
Regarding syntax errors, it has been tested and there are no errors. 

I think the problem may be in the condition for the playbook to be executed. If I wanted the playbook to run only after a certain number of attempts, what would that condition be? I read somewhere that the count is deprecated. I tried with 

```
threshold:
  field: "@timestamp"
   timeframe: 1h
   value: 6
```
And now with 

```
occurrences:
   from: 6
   to: 999999
```
And still no playbook alert. :/

**Reply by keis3cker** [reply]
I would test basic functionality first. And when this works going further.
You can also look at the elastalert logs, to see if any error occures at execution time.

**Reply by Mav1814** [reply]
I tested the basic functionality, the output of the sigma rule to elasticsearch query works if i put her in kibana.
This is my detection on the sigma side.
```
detection:
  condition: selection
  selection:
    - message|endswith: "Authentication Failure"
    - message|contains: "Invalid User"
```

And this is the output to elasticsearch.

```
any where (message : "*Authentication Failure" or message : "*Invalid User*")
```

 I will try to check the logs to clues.

**Reply by Mav1814** [reply]
UPDATE:

The playbook is working without any error. As seen in the image bellow
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/7c1c0fa7-1256-4311-b8a1-49e0b206e8e2)

No events, or detection. So from now i know that the problem is on sigma rule. So i need help (or guides) to understand how to create a rule for Authentication Failure. I've been searching but i didn't find anything in concrete to help me.

**Reply by Mav1814** [reply]
Hi again!

I've been very frustrated by the playbooks. I've seen and reviewed the sigma rules documentation, I've done countless tests, all of them without success.  Can anyone help me? Or tell me what's missing?
The playbook runs successfully, the elastic query works perfectly. Only no alert is generated. 

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/1a4f74fe-0b19-4cd4-b16a-e5381576d70b)


I've tried in logsource, with service: auth, sshd, etc. Supposedly without service, just with the product for linux, it should fetch any log where the keyword has "Authentication Failure". I've also tried searching directly for "message: "authentication failure"", without success. When the rules are converted to the elastic query, it always works if I search directly in kibana. I can't understand why it doesn't generate an alert.
I've seen countless examples in Linux, I don't understand what's failing.

**Reply by Mav1814** [reply]
I was able to make a windows playbook work, on logs it shows 
```
2023-10-18 14:13:20,809     INFO           elastalert Ran Failed Windows Login - 34fd55z4d from 2023-10-18 14:03 UTC to 2023-10-18 14:13 UTC: 4 query hits (0 already seen), 4 matches, 4 alerts sent
```
But on Alerts side, no alert was create, why?? Any suggestion? 

And sorry about the spam, but im really frustrated.

**Reply by jaypfe** [reply]
In Kibana, go to discover. On the left hand side where the data views are at, select 'create a data view'
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/8bece47a-a9b8-4192-b2ff-5b6a8667b6c2)

Call it whatever and use the index pattern shown:
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/931d5584-0c46-438c-9810-8c1417473100)

This is were the playbook alerts get index when they are fired. If you see anything here, you should see it in the alerts page as long as your time frame is the same. 

If you don't see anything here, try looking at the elastalert_status view in discover and see if you see the rules being ran. 

You can also look in /opt/so/rules/elastalert/playbook and make sure a rule was made. It will have a name of random characters. 

also also, make sure your play is active of course.

**Reply by Mav1814** [reply]
Hello @jaypfe 

I really appreciate your help.  I've already followed the steps you told me, but I couldn't find this kind of information in the documentation. 
Responding to what you wrote. 

1 - elastalert_status - I actually see that the rule has run
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/42f3ba69-34c4-4bbe-8f5c-853d26b6a8b0)

2 - After creating the custom index and trying the playbook again, i have no results by searching for the rule.name : "Failed Windows Login". (but others playbook rules.name apear)

3 - In the /opt/so/rules/elastalert/playbook logs, I see 

```
2023-10-19 09:14:40,198     INFO apscheduler.executors.default Running job "Rule: Failed Windows Login - 34fd55z4d (trigger: interval[0:03:00], next run at: 2023-10-19 09:17:43 UTC)" (scheduled at 2023-10-19 09:14:40.197661+00:00)
2023-10-19 09:14:40,367     INFO           elastalert Queried rule Failed Windows Login - 34fd55z4d from 2023-10-19 09:04 UTC to 2023-10-19 09:14 UTC: 6 / 6 hits
2023-10-19 09:14:40,536     INFO           elastalert Ran Failed Windows Login - 34fd55z4d from 2023-10-19 09:04 UTC to 2023-10-19 09:14 UTC: 6 query hits (0 already seen), 6 matches, 6 alerts sent
2023-10-19 09:14:40,536     INFO           elastalert Failed Windows Login - 34fd55z4d range 600
```

In fact, 6 hits appear, and an alert is sent, but nothing appears on the kibana side of the index. 

I noticed that in the specific windows log, this information doesn't appear (another linux log) could this be the case? if so, how do I solve it? 

![log](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/4cbbe929-11ef-40c2-8aa0-cb65a9ff4869)

I can't really find out what's going wrong.

**Reply by jaypfe** [reply]
So just to be clear, you can see the alerts pop up in the alerts section of the SOC? Just not in Kibana? 

Sorry I am a bit confused on exactly what is happening. Is the problem limited to this logon rule? You have created other plays that work as expected?

**Reply by Mav1814** [reply]
I'll try to explain, I apologize for the lack of clarity.

Initially I was trying to create a playbook for Linux that could generate alerts for failed login attempts. Without success, I'm not succeeding, although it doesn't give any error, it doesn't generate any alert, nor is it trigered. I still don't know why.
So I tried to create a playbook for windows, to generate the same type of alerts, in which case, in the logs in /opt/so/rules/elastalert/playbook as you can see above, it detects and supposedly sends the alert. But no alert reaches the SO. 
So I imported the playbooks from the repository of available sigma rules (windows, linux, cloud, macOS...) and ran a test on linux to try to understand the behavior. I tested it with the "Linux Uninstall Package" playbook and everything worked normally, the alert was generated, both on the kibanam side and on the SO side. I used a linux playbook, because it was more easy for me to test.
What I realized when comparing the two logs  on kibana (from linux and windows) was that:
1 - Linux - _index: .ds-logs-playbook.alerts-so-2023.10.06-000001
2 - Windows - _index: .ds-logs-system.security-default-2023.10.15-000002
On linux it still shows event_data._index: .ds-logs-endpoint.events.process-default-2023.10.15-000002
On the windows side, nothing appears.
It feels like it's not converting or something, I don't know what's going on. I can post the playbook here.

**Reply by jaypfe** [reply]
Yes please post the playbook. Also the path you are referencing should not have any logs in it. 

Please make sure you are going to: /opt/so/RULES/elastalert/playbook. You should see something like this:
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/d629da74-ab58-42b6-bf81-08eb50782ed7)

It can be hard to find the rule in question here because they are named with just numbers. But you can cat them and take a look at the name within the rule.  We want to look in that rule, and directly compare it to the one you know works. Based on what you are saying here, I would expect its looking in the wrong log source. If your Linux rule is searching in playbook alerts, that is probably it right there. 

Also, when you find the rule that is not working, run a sudo so-elastalert-test and select that rule. That script gives you a pretty good idea of whats happening. Post the output here if you would.

**Reply by jaypfe** [reply]
You said you have tested the EQL? It works as designed?

**Reply by Mav1814** [reply]
Yes, I can find my playbook there /opt/so/RULES/elastalert/playbook. I even compared it with others playbooks to see if there were any differences. I'll post the playbook here. I didnt know about so-elastalert-test.

**Reply by Mav1814** [reply]
This is the playbook
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/793e928c-6cf8-4103-9cbf-9cce27084e5f)

**Reply by Mav1814** [reply]
And this is the other part of the playbook configuration.

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/cc05da7c-b997-4ce5-b188-9600e95c47cb)

**Reply by Mav1814** [reply]
And this is the file on /opt/so/RULES/elastalert/playbook

![log](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/5ba26918-84fc-4d9a-9fed-562de153a111)

**Reply by Mav1814** [reply]
And the output of so-elastalert-test 72de8056a.yaml 

```
Choose a rule to test from the list above (must be typed exactly as shown above): playbook/72de8056a.yaml
The results can be rather long.  Would you like to write the results to a file? (y/N) N

Got N/A hits from the last 0 day

Available terms in first hit:
	agent.name
	agent.id
	agent.ephemeral_id
	agent.type
	agent.version
	metadata.input.beats.host.ip
	metadata.raw_index
	metadata.stream_id
	metadata.beat
	metadata.type
	metadata.version
	metadata.input_id
	process.name
	process.pid
	process.executable
	winlog.computer_name
	winlog.process.pid
	winlog.process.thread.id
	winlog.keywords
	winlog.logon.failure.reason
	winlog.logon.failure.sub_status
	winlog.logon.failure.status
	winlog.logon.id
	winlog.logon.type
	winlog.channel
	winlog.event_data.Status
	winlog.event_data.LogonType
	winlog.event_data.IpPort
	winlog.event_data.TransmittedServices
	winlog.event_data.SubjectLogonId
	winlog.event_data.LmPackageName
	winlog.event_data.KeyLength
	winlog.event_data.SubjectUserName
	winlog.event_data.FailureReason
	winlog.event_data.SubjectDomainName
	winlog.event_data.IpAddress
	winlog.event_data.TargetUserName
	winlog.event_data.SubStatus
	winlog.event_data.TargetDomainName
	winlog.event_data.LogonProcessName
	winlog.event_data.SubjectUserSid
	winlog.event_data.TargetUserSid
	winlog.event_data.AuthenticationPackageName
	winlog.opcode
	winlog.record_id
	winlog.task
	winlog.event_id
	winlog.provider_guid
	winlog.activity_id
	winlog.api
	winlog.provider_name
	log.level
	elastic_agent.id
	elastic_agent.version
	elastic_agent.snapshot
	source.domain
	message
	type
	error.message
	tags
	module_temp
	dataset_tag_temp
	input.type
	@timestamp
	ecs.version
	related.user
	data_stream.namespace
	data_stream.type
	data_stream.dataset
	@version
	host.hostname
	host.os.build
	host.os.kernel
	host.os.name
	host.os.type
	host.os.family
	host.os.version
	host.os.platform
	host.ip
	host.name
	host.id
	host.mac
	host.architecture
	event.agent_id_status
	event.ingested
	event.code
	event.provider
	event.kind
	event.created
	event.module
	event.action
	event.category
	event.type
	event.dataset
	event.outcome
	user.domain
	user.name
	user.id


Would have written the following documents to writeback index (default is elastalert_status):

elastalert_status - {'rule_name': 'Failed Windows Login - 72de8056a', 'endtime': datetime.datetime(2023, 10, 19, 14, 28, 33, 737737, tzinfo=tzutc()), 'starttime': datetime.datetime(2023, 10, 18, 14, 28, 33, 737737, tzinfo=tzutc()), 'matches': 18, 'hits': 18, '@timestamp': datetime.datetime(2023, 10, 19, 14, 28, 57, 625388, tzinfo=tzutc()), 'time_taken': 23.513360023498535}


Test completed successfully!
```

**Reply by jaypfe** [reply]
I just made one that functions as it should. I think you need to remove your 'status' Draft from your play. 
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/217feef3-a7a2-4706-bc58-63d94bdecc03)
Change it to experimental.

**Reply by Mav1814** [reply]
I didn't put the status as experimental because on the documentation they say " You may also want to avoid others with a status of experimental."

I will test your playbook.

**Reply by Mav1814** [reply]
I tested, same result. :( 
so if for you it generate a alert, the problem is not the playbook itself. I dont know how to fix this.

**Reply by jaypfe** [reply]
The play is set to active? A new Elastalert rule was built in /opt/so/rule? 
If you test the EQL do you get hits? 
Can you see that the rule is running? 

Any errors when running a highstate? 
You may try restarting the grid. 

You dont have any custom pipes or data streams do you? 

Beyond this, I think one of the onion lords themselves would have to look at it. They would probably know immediately.

**Reply by Mav1814** [reply]
Hello!
The behavior is exactly the same as I had with the previous playbook. 
Answers

Is the play set to active? 
A: yes

Has a new Elastalert rule been created in /opt/so/rule?
A: yes

If you test EQL, do you get results?
A: yes

Can you see that the rule is being executed?
A: yes, and is triggered and "alert sent"

Any errors when running a highstate?
You may try restarting the grid.
A: didn't try yet.

You dont have any custom pipes or data streams do you?
A: nop, nothing. I only have 2 endpoints with agents. I didn't change or customize anything yet. 

Beyond this, I think one of the onion lords themselves would have to look at it. They would probably know immediately.
A: I wish :/

Much thanks for everything.

PS: I feel like I've been wasting a lot of time understanding security onion, trying to explore its features and trying to understand how it can be useful, but all I find are difficulties and unsolved problems. It's been quite frustrating, I've been doing this for a long time. 
As much as I try to dig deeper, it becomes difficult, I think I've read the available documentation a couple of times, often without an answer. I even understand that it's natural, since there's a free part and a paid part, but... I think this solution has a lot of potential and can solve a lot of problems, and for sure thats an incredible work, but I've really found it difficult to exploit. It was just an aside, I'm sorry.

**Reply by Mav1814** [reply]
@jaypfe Can you show me the configuration that u did when creating the play? What did you put on "Group"?

**Comment by jaypfe** [comment]
Sure, I did not fill any of those fields out in my example.

If you can download community plays and they work as they should, it HAS to be something that is either being done/ not being done in the creation. 
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/1ef82175-8a71-402a-8fc9-7bb2d33b641c)

**Reply by Mav1814** [reply]
@jaypfe yep, i did a test, put one of the playbooks in the community as inactive, and created it manually and it worked normally. So the problem could be down to configuration.

I'll test it as you did and see if I can see any changes.

**Reply by Mav1814** [reply]
The same result. 

```
2023-10-20 15:22:20,879     INFO           elastalert Ran Failed Windows Login - 72de8056a from 2023-10-20 15:18 UTC to 2023-10-20 15:22 UTC: 3 query hits (0 already seen), 3 matches, 3 alerts sent
```
but it doesn't send any alert to SO.

@jaypfe Can you show the content on the playID.yaml please?

**Reply by jaypfe** [reply]
sure: 
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/fc980e2c-3448-4289-9a87-5c2e7aaaa49a)

**Reply by jaypfe** [reply]
I would urge you run a high state, and try restarting the grid. 

Especially if you have not restarted since you built it.

**Reply by Mav1814** [reply]
its the same, i can't figure out what is happening here.. why isn't working on my side. 

I will restart everything. I've a distributed deployment.

@jaypfe I'm running sudo salt-call state.highstate right now. Then i will reboot.

**Reply by jaypfe** [reply]
Run this command after restarting and everything comes back online. 
sudo salt \* state.highstate

See if there are any errors. 

Also how much ram for each node in your deployment?

**Reply by Mav1814** [reply]
After running  sudo salt * state.highstate
```
No minions matched the target. No command was sent, no jid was assigned.
ERROR: No return received
```
They have 32GB RAM each.

**Reply by jaypfe** [reply]
You ran that on the manager right?

**Reply by Mav1814** [reply]
right :) 

I dont know if is even possible on the nodes. xD

**Reply by jaypfe** [reply]
After a restart, try running it again. 
Beyond this man, I am out of ideas.

**Reply by Mav1814** [reply]
I already appreciate everything that you did/try to help me! thank you is not enough even this is not working! At least you tried.

I will reboot everything know, then i will running it again. But the output is okay right? its suppose to show this output?

**Reply by jaypfe** [reply]
No that is not what should happen. It will give a large amount of output for each container running on the nodes. It takes a bit of time to complete. 

That is also why when creating the post it asks you to run it, as it can give you good clues as to what is happening.

**Reply by Mav1814** [reply]
Sorry, i missed that :/.

Everything rebooted and up and running. My questions are

1- If i run sudo salt * state.highstate the output is the same as above.
2 - if i run sudo  salt-call * state.highstate it says that 'Security Onion is not available', (dont know why it never happened) so i logged as root xD 
The final output after running in root was:

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/165395f5-6cbf-4e1c-8a4b-0fc4e802e7d9)

I dont know if its the same command or not.

**Reply by jaypfe** [reply]
That looks more like the output you should expect. Honestly, I am not sure on the command differences. 
But just for grins, does the play work now?

**Reply by Mav1814** [reply]
Hi @jaypfe 

The result is the same. 

```
2023-10-23 10:18:12,295     INFO apscheduler.executors.default Running job "Rule: Failed Windows Login - 72de8056a (trigger: interval[0:03:00], next run at: 2023-10-23 10:21:16 UTC)" (scheduled at 2023-10-23 10:18:12.294521+00:00)
2023-10-23 10:18:12,504     INFO           elastalert Queried rule Failed Windows Login - 72de8056a from 2023-10-23 10:08 UTC to 2023-10-23 10:18 UTC: 3 / 3 hits
2023-10-23 10:18:12,695     INFO           elastalert Ran Failed Windows Login - 72de8056a from 2023-10-23 10:08 UTC to 2023-10-23 10:18 UTC: 3 query hits (0 already seen), 3 matches, 3 alerts sent
2023-10-23 10:18:12,695     INFO           elastalert Failed Windows Login - 72de8056a range 600
2023-10-23 10:18:12,695     INFO apscheduler.executors.default Job "Rule: Failed Windows Login - 72de8056a (trigger: interval[0:03:00], next run at: 2023-10-23 10:21:16 UTC)" executed successfully
```
And no alert is sent to Alerts on SO.

**Reply by Mav1814** [reply]
@jaypfe I did a reset on the playbooks. I can't re-import any playbook now, not even from sigma or the IDH.  xD

But i tried to recreate the windows play for auth failed, and know it works, i don't know why. It's a bit weird, but know is working. :)

I have a question, i'm trying to do the same for linux, but no success, not even a match. Can you help me ? Sorry to bother you again.
